src/main/java/hudson/plugins/ec2/SlaveTemplate.java

```diff
@@ -1164,44 +1167,77 @@ private List<EC2AbstractSlave> provisionOndemand(Image image, int number, EnumSe
+
             try {
-                newInstances = ec2.runInstances(riRequest).getReservation().getInstances();
+                newInstances = createNewInstances(ec2, riRequest);
             } catch (AmazonEC2Exception e) {
                 if (fallbackSpotToOndemand && e.getErrorCode().equals("InsufficientInstanceCapacity")) {
                     logProvisionInfo("There is no spot capacity available matching your request, falling back to on-demand instance.");
                     riRequest.setInstanceMarketOptions(new InstanceMarketOptionsRequest());
-                    newInstances = ec2.runInstances(riRequest).getReservation().getInstances();
+                    newInstances = createNewInstances(ec2, riRequest);
                 } else {
                     throw e;
                 }
             }
         } else {
-            newInstances = ec2.runInstances(riRequest).getReservation().getInstances();
+            newInstances = createNewInstances(ec2, riRequest);
         }
         // Have to create a new instance
 
         if (newInstances.isEmpty()) {
             logProvisionInfo("No new instances were created");
         }
 
-        newInstances.addAll(orphansOrStopped);
+        newInstances.addAll(InstanceInfo.fromInstances(orphansOrStopped, getParent()));
 
         return toSlaves(newInstances);
     }
 
+    List<InstanceInfo> createNewInstances(AmazonEC2 ec2, RunInstancesRequest riRequest) {
+        if (isUsingDynamicSshKeys()) {
+            // no static ssh key defined, so submit riRequest.maxCount requests
+            // each with a unique keypair (instead of a single request for multiple identical instances)
+            LOGGER.fine(() -> "no static ssh credential configured, will use dynamic ssh keys instead");
+            List<InstanceInfo> instances = new ArrayList<>();
+            int maxRequested = riRequest.getMaxCount();
+            riRequest.setMaxCount(1);
+            for (int i=0;i < maxRequested; i++) {
+                String keyName = "cloudbees-" + UUID.randomUUID();
+                KeyPair keyPair = ec2.createKeyPair(new CreateKeyPairRequest(keyName)).getKeyPair();
+                LOGGER.fine("created new dynamic keypair " + keyPair.getKeyName());
+                riRequest.setKeyName(keyPair.getKeyName());
+                Instance instance = ec2.runInstances(riRequest).getReservation().getInstances().get(0);
+                instances.add(new InstanceInfo(instance, keyPair));
+            }
+            return instances;
+        } else {
+            // using a static ssh key
+            List<InstanceInfo> instances = InstanceInfo.fromInstances(ec2.runInstances(riRequest).getReservation().getInstances(), getParent());
+            instances.stream().forEach(instance -> {
+                try {
+                    LOGGER.fine(() -> "static ssh credential configured, retrieving keypair and setting it in instance");
+                    instance.setKeypair(getParent().resolveKeyPair());
```

Comment:
Why was https://github.com/jenkinsci/ec2-plugin/pull/993#discussion_r1822857034 marked resolved?

Reply:
Apparently i missed removing that after implementing the requested change :facepalm:  see 
[2bec189](https://github.com/jenkinsci/ec2-plugin/pull/993/commits/2bec189d50906915f549eb2012e65cf3e85b08b5)